File: platform/org-ops/points/scoring.ts
/**
 * Point scoring — turns the append-only ledger into derived per-member standings.
 *
 * Weights encode the confirmed priority: achievements/grades/projects are the highest-priority
 * signals; referral is the lightest. Weights are the ENGINE DEFAULT; the SQL stores a `weight`
 * per row so admins can override per event. When an event carries its own weight, we trust it;
 * otherwise we apply the source default below.
 */
import type { MemberStanding, PointEvent, PointSource } from "./types";

/** Default per-source weight. Tunable; mirrors the intent in docs/POINTS-ENGINE.md. */
export const DEFAULT_SOURCE_WEIGHTS: Readonly<Record<PointSource, number>> = {
  achievement: 5,
  grade: 4,
  project: 4,
  activity: 2,
  referral: 1,
};

/** The weight to use for an event: its explicit weight if positive, else the source default. */
export function effectiveWeight(event: PointEvent): number {
  return event.weight > 0 ? event.weight : DEFAULT_SOURCE_WEIGHTS[event.source];
}

/** Weighted contribution of a single ledger row. */
export function weightedPoints(event: PointEvent): number {
  return event.points * effectiveWeight(event);
}

function minIso(a: string | null, b: string): string {
  return a === null || b < a ? b : a;
}

function maxIso(a: string | null, b: string): string {
  return a === null || b > a ? b : a;
}

/**
 * Fold the ledger into one standing per member. `nicknames` supplies the public-safe label;
 * a member with no nickname falls back to their id so the engine never invents PII.
 */
export function aggregateStandings(
  events: readonly PointEvent[],
  nicknames: ReadonlyMap<string, string>,
): MemberStanding[] {
  const byMember = new Map<string, MemberStanding>();

  for (const event of events) {
    const existing = byMember.get(event.memberId);
    const contribution = weightedPoints(event);
    if (existing) {
      byMember.set(event.memberId, {
        ...existing,
        totalPoints: existing.totalPoints + contribution,
        firstEarnedAt: minIso(existing.firstEarnedAt, event.earnedAt),
        lastActiveAt: maxIso(existing.lastActiveAt, event.earnedAt),
      });
    } else {
      byMember.set(event.memberId, {
        memberId: event.memberId,
        totalPoints: contribution,
        firstEarnedAt: event.earnedAt,
        lastActiveAt: event.earnedAt,
        nickname: nicknames.get(event.memberId) ?? event.memberId,
      });
    }
  }

  return [...byMember.values()];
}

File: src/org_ops/points/scoring.py
"""Point scoring — turns the append-only ledger into derived per-member standings.

Weights encode the confirmed priority: achievements/grades/projects are the highest-priority
signals; referral is the lightest. Weights are the ENGINE DEFAULT; the SQL stores a `weight`
per row so admins can override per event. When an event carries its own weight, we trust it;
otherwise we apply the source default below.

Python port of platform/org-ops/points/scoring.ts.
"""

from __future__ import annotations

from org_ops.points.types import MemberStanding, PointEvent, PointSource

# Default per-source weight. Tunable; mirrors the intent in docs/POINTS-ENGINE.md.
DEFAULT_SOURCE_WEIGHTS: dict[str, float] = {
    PointSource.ACHIEVEMENT: 5,
    PointSource.GRADE: 4,
    PointSource.PROJECT: 4,
    PointSource.ACTIVITY: 2,
    PointSource.REFERRAL: 1,
}


def effective_weight(event: PointEvent) -> float:
    """The weight to use for an event: its explicit weight if positive, else the source default."""
    return event.weight if event.weight > 0 else DEFAULT_SOURCE_WEIGHTS[event.source]


def weighted_points(event: PointEvent) -> float:
    """Weighted contribution of a single ledger row."""
    return event.points * effective_weight(event)


def _min_iso(a: str | None, b: str) -> str:
    """Return the earlier of two ISO timestamps; treats None as 'no earlier bound'."""
    return b if a is None or b < a else a


def _max_iso(a: str | None, b: str) -> str:
    """Return the later of two ISO timestamps; treats None as 'no later bound'."""
    return b if a is None or b > a else a


def aggregate_standings(
    events: list[PointEvent],
    nicknames: dict[str, str],
) -> list[MemberStanding]:
    """Fold the ledger into one standing per member.

    ``nicknames`` supplies the public-safe label; a member with no nickname falls back to their
    id so the engine never invents PII.

    Returns a new list; the input ``events`` sequence is never mutated.
    """
    by_member: dict[str, MemberStanding] = {}

    for event in events:
        contribution = weighted_points(event)
        if event.member_id in by_member:
            existing = by_member[event.member_id]
            by_member[event.member_id] = existing.model_copy(
                update={
                    "total_points": existing.total_points + contribution,
                    "first_earned_at": _min_iso(existing.first_earned_at, event.earned_at),
                    "last_active_at": _max_iso(existing.last_active_at, event.earned_at),
                }
            )
        else:
            by_member[event.member_id] = MemberStanding(
                member_id=event.member_id,
                total_points=contribution,
                first_earned_at=event.earned_at,
                last_active_at=event.earned_at,
                nickname=nicknames.get(event.member_id, event.member_id),
            )

    return list(by_member.values())
